catch2: Upgrade to v3.1.0

mikecrowe opened the pull request:
Catch v3 now builds as a library rather than a single include file, which means that the meson.build files need to be rather more complex.

Catch v3's own CMake build system doesn't build dynamic libraries, and it doesn't seem like doing so would be worthwhile for such a project, so the Meson build only builds static libraries.

neheb commented:
Whoops. Meant to just run the CI

eli-schwartz commented on subprojects/packagefiles/catch2/src/catch2/meson.build:
```diff
@@ -0,0 +1,338 @@
+  'Catch2Main',
+  [ 'internal/catch_main.cpp' ],
+  include_directories : '..',
+  dependencies : [],
```
You should add `link_with: catch2`, in order to guarantee that dependencies are traced.

eli-schwartz commented on subprojects/packagefiles/catch2/src/catch2/meson.build:
```diff
@@ -0,0 +1,338 @@
+  libraries : catch2_with_main,
+  version : meson.project_version(),
+  description : 'A modern, C++-native, test framework for C++14 and above (links in default main)',
+  requires : 'catch2 = ' + meson.project_version()
```
And here, the link_with will cause pkg.generate to detect that catch2_with_main depends on the "catch2" pkg-config dependency, although this automatic handling doesn't add a version match, hmm...

eli-schwartz commented on subprojects/packagefiles/catch2/meson.build:
```diff
@@ -1,9 +1,10 @@
+  version : '3.1.0',
+  license: 'Boost 1.0')
+pkg = import('pkgconfig')
+cpp = meson.get_compiler('cpp')
```
As far as I can tell you aren't using the found compiler for anything?

eli-schwartz commented on subprojects/packagefiles/catch2/meson.build:
```diff
@@ -1,9 +1,10 @@
+  'catch2',
+  'cpp',
+  version : '3.1.0',
+  license: 'Boost 1.0')
```
Once we are updating the license to make it more accurate, we might as well use the SPDX identifier,  `BSL-1.0`.

eli-schwartz commented on subprojects/packagefiles/catch2/src/catch2/meson.build:
```diff
@@ -0,0 +1,338 @@
+    folder = ''
+    foreach path : file_path
+        if path != file_path[-1]
+            folder = folder / path
```
This is currently the used feature with the highest minimum version of Meson, at a pretty backwards compatible 0.49.0 -- great stuff.

Can we assert that in code by adding `meson_version: '>=0.49.0'` in the `project()` function kwargs?

eli-schwartz reviewed (dismissed):
The CI is failing because this no longer provides `catch2_dep = declare_dependency(...)` -- it would be apropos to provide one using `link_with: catch2, include_directories: 'src'`.

Now that catch2-with-main.pc is an upstream thing, I think it would be nice to additionally add that to the `[provide]` section of the wrap file.

eli-schwartz commented:
> Catch v3's own CMake build system doesn't build dynamic libraries, and it doesn't seem like doing so would be worthwhile for such a project, so the Meson build only builds static libraries.

Excellent analysis, I agree.

mikecrowe commented:
Thank you for the comprehensive and positive review. I believe that I've addressed all your points. I did need to use `include_directories: '..'` when declaring the dependencies rather than `include_directories: 'src'` though. Once these, and any future fixes, have been reviewed I can squash them down to a single commit.

I was also able to actually test the results rather better this time too by copying the wrap file and symlinking my Catch repository into another project that used Catch2. Is there a better way to do this that I've missed mention in the documentation for?

eli-schwartz commented:
> I did need to use `include_directories: '..'` when declaring the dependencies rather than `include_directories: 'src'` though.

Yeah, that's fine, it just depends on which directory you create it from. :)



> I was also able to actually test the results rather better this time too by copying the wrap file and symlinking my Catch repository into another project that used Catch2. Is there a better way to do this that I've missed mention in the documentation for?

The wrapdb repository itself can be used, via `meson setup -Dwraps=catch2 builddir/`, and that's how we test it in CI. That won't test using it in a project that makes use of catch2 though. For that, the usual real solution is, in fact, to copy over the .wrap file and e.g. the subprojects/packagefiles/catch2/ directory or the extracted sources.

The other solution is to run `python3 tools/create_release.py` which when run without authentication tokens for https://wrapdb.mesonbuild.com will simply generate catch2_3.1.0-1_patch.zip and catch2.wrap into subprojects/packagecache/

This is not precisely documented, though.

eli-schwartz commented:
There are two CI failures:
- one of them is irrelevant and just means you need to rebase after another PR was merged
- the other is saying that the catch2-with-main dependency isn't mentioned in releases.json

mikecrowe commented:
I added `"catch2-with-main"` to `dependency_names` in `release.json` as you suggested, rebased, and it looks like the tests now pass. Thanks for that.

Whilst reviewing the patch I did realise that I've been rather inconsistent in where I've put the closing parentheses. I can make them consistent, but to do that I've to decide whether they should be on a line on their own for multi-line function invocations (e.g. as currently used for `pkg.generate`) or at the end of the last line (e.g. as currently used for `static_library`). Which do you prefer? Looking at the existing files there seems to be a preference for the former, though there are many instances of the latter too.

Thanks.

eli-schwartz commented:
FWIW my personal preference tends to the former.

mikecrowe commented:
eli-shwartz wrote:
> FWIW my personal preference tends to [parentheses on their own line].

After initially managing to do exactly the wrong thing, I believe that I've now done this correctly. I also switched to consistently using two spaces for indents. (Previously, for some reason Emacs wanted to use two characters for the function invocations but four characters for the `foreach`.) Please let me know if you'd like me to change anything else.

eli-schwartz commented:
Thanks! Looks great.

mikecrowe commented:
Thank you for the reviews and for merging the change.

neheb commented:
This PR breaks cppzmq:

```
../subprojects/cppzmq-4.8.1/tests/recv_multipart.cpp:1:10: fatal error: catch2/catch.hpp: No such file or directory
    1 | #include <catch2/catch.hpp>
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.
[8/14] Compiling C++ object subprojects/cppzmq-4.8.1/unit_tests.p/tests_socket.cpp.o
FAILED: subprojects/cppzmq-4.8.1/unit_tests.p/tests_socket.cpp.o 
ccache c++ -Isubprojects/cppzmq-4.8.1/unit_tests.p -Isubprojects/cppzmq-4.8.1 -I../subprojects/cppzmq-4.8.1 -I../subprojects/cppzmq-4.8.1/tests -Isubprojects/Catch2-3.1.0/src -I../subprojects/Catch2-3.1.0/src -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++17 -O0 -g -pthread -MD -MQ subprojects/cppzmq-4.8.1/unit_tests.p/tests_socket.cpp.o -MF subprojects/cppzmq-4.8.1/unit_tests.p/tests_socket.cpp.o.d -o subprojects/cppzmq-4.8.1/unit_tests.p/tests_socket.cpp.o -c ../subprojects/cppzmq-4.8.1/tests/socket.cpp
../subprojects/cppzmq-4.8.1/tests/socket.cpp:1:10: fatal error: catch2/catch.hpp: No such file or directory
    1 | #include <catch2/catch.hpp>
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.
[9/14] Compiling C++ object subprojects/cppzmq-4.8.1/unit_tests.p/tests_message.cpp.o
FAILED: subprojects/cppzmq-4.8.1/unit_tests.p/tests_message.cpp.o 
ccache c++ -Isubprojects/cppzmq-4.8.1/unit_tests.p -Isubprojects/cppzmq-4.8.1 -I../subprojects/cppzmq-4.8.1 -I../subprojects/cppzmq-4.8.1/tests -Isubprojects/Catch2-3.1.0/src -I../subprojects/Catch2-3.1.0/src -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++17 -O0 -g -pthread -MD -MQ subprojects/cppzmq-4.8.1/unit_tests.p/tests_message.cpp.o -MF subprojects/cppzmq-4.8.1/unit_tests.p/tests_message.cpp.o.d -o subprojects/cppzmq-4.8.1/unit_tests.p/tests_message.cpp.o -c ../subprojects/cppzmq-4.8.1/tests/message.cpp
../subprojects/cppzmq-4.8.1/tests/message.cpp:2:10: fatal error: catch2/catch.hpp: No such file or directory
    2 | #include <catch2/catch.hpp>
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.
[10/14] Compiling C++ object subprojects/cppzmq-4.8.1/unit_tests.p/tests_context.cpp.o
FAILED: subprojects/cppzmq-4.8.1/unit_tests.p/tests_context.cpp.o 
ccache c++ -Isubprojects/cppzmq-4.8.1/unit_tests.p -Isubprojects/cppzmq-4.8.1 -I../subprojects/cppzmq-4.8.1 -I../subprojects/cppzmq-4.8.1/tests -Isubprojects/Catch2-3.1.0/src -I../subprojects/Catch2-3.1.0/src -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++17 -O0 -g -pthread -MD -MQ subprojects/cppzmq-4.8.1/unit_tests.p/tests_context.cpp.o -MF subprojects/cppzmq-4.8.1/unit_tests.p/tests_context.cpp.o.d -o subprojects/cppzmq-4.8.1/unit_tests.p/tests_context.cpp.o -c ../subprojects/cppzmq-4.8.1/tests/context.cpp
../subprojects/cppzmq-4.8.1/tests/context.cpp:1:10: fatal error: catch2/catch.hpp: No such file or directory
    1 | #include <catch2/catch.hpp>
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.
[11/14] Compiling C++ object subprojects/cppzmq-4.8.1/unit_tests.p/tests_monitor.cpp.o
FAILED: subprojects/cppzmq-4.8.1/unit_tests.p/tests_monitor.cpp.o 
ccache c++ -Isubprojects/cppzmq-4.8.1/unit_tests.p -Isubprojects/cppzmq-4.8.1 -I../subprojects/cppzmq-4.8.1 -I../subprojects/cppzmq-4.8.1/tests -Isubprojects/Catch2-3.1.0/src -I../subprojects/Catch2-3.1.0/src -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++17 -O0 -g -pthread -MD -MQ subprojects/cppzmq-4.8.1/unit_tests.p/tests_monitor.cpp.o -MF subprojects/cppzmq-4.8.1/unit_tests.p/tests_monitor.cpp.o.d -o subprojects/cppzmq-4.8.1/unit_tests.p/tests_monitor.cpp.o -c ../subprojects/cppzmq-4.8.1/tests/monitor.cpp
In file included from ../subprojects/cppzmq-4.8.1/tests/monitor.cpp:1:
../subprojects/cppzmq-4.8.1/tests/testutil.hpp:3:10: fatal error: catch2/catch.hpp: No such file or directory
    3 | #include <catch2/catch.hpp>
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.
[12/14] Compiling C++ object subprojects/cppzmq-4.8.1/unit_tests.p/tests_buffer.cpp.o
FAILED: subprojects/cppzmq-4.8.1/unit_tests.p/tests_buffer.cpp.o 
ccache c++ -Isubprojects/cppzmq-4.8.1/unit_tests.p -Isubprojects/cppzmq-4.8.1 -I../subprojects/cppzmq-4.8.1 -I../subprojects/cppzmq-4.8.1/tests -Isubprojects/Catch2-3.1.0/src -I../subprojects/Catch2-3.1.0/src -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++17 -O0 -g -pthread -MD -MQ subprojects/cppzmq-4.8.1/unit_tests.p/tests_buffer.cpp.o -MF subprojects/cppzmq-4.8.1/unit_tests.p/tests_buffer.cpp.o.d -o subprojects/cppzmq-4.8.1/unit_tests.p/tests_buffer.cpp.o -c ../subprojects/cppzmq-4.8.1/tests/buffer.cpp
../subprojects/cppzmq-4.8.1/tests/buffer.cpp:1:10: fatal error: catch2/catch.hpp: No such file or directory
    1 | #include <catch2/catch.hpp>
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.
[13/14] Compiling C++ object subprojects/cppzmq-4.8.1/unit_tests.p/tests_active_poller.cpp.o
FAILED: subprojects/cppzmq-4.8.1/unit_tests.p/tests_active_poller.cpp.o 
ccache c++ -Isubprojects/cppzmq-4.8.1/unit_tests.p -Isubprojects/cppzmq-4.8.1 -I../subprojects/cppzmq-4.8.1 -I../subprojects/cppzmq-4.8.1/tests -Isubprojects/Catch2-3.1.0/src -I../subprojects/Catch2-3.1.0/src -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++17 -O0 -g -pthread -MD -MQ subprojects/cppzmq-4.8.1/unit_tests.p/tests_active_poller.cpp.o -MF subprojects/cppzmq-4.8.1/unit_tests.p/tests_active_poller.cpp.o.d -o subprojects/cppzmq-4.8.1/unit_tests.p/tests_active_poller.cpp.o -c ../subprojects/cppzmq-4.8.1/tests/active_poller.cpp
In file included from ../subprojects/cppzmq-4.8.1/tests/active_poller.cpp:3:
../subprojects/cppzmq-4.8.1/tests/testutil.hpp:3:10: fatal error: catch2/catch.hpp: No such file or directory
    3 | #include <catch2/catch.hpp>
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.
ninja: build stopped: subcommand failed.
```

eli-schwartz commented:
So "catch2" updating from version 2 to version 3 is incompatible, which is an interesting naming scheme but okay.

mikecrowe commented:
It's a bit confusing, but ["Catch2" is now the name of the project](https://levelofindirection.com/blog/catch2-released.html), not its version number. See also: https://github.com/catchorg/Catch2/issues/769

The project tries to follow [semantic versioning](https://github.com/catchorg/Catch2/blob/devel/docs/release-process.md), so Catch2 v3.1.0 isn't API compatible with Catch2 v2.x.

Does this mean that we need separate catch2-2 and catch2-3 in the wrapdb? Or should we just wait for cppzmq to catch up?

neheb commented:
Probably the latter. I don't know if this was fixed upstream.

eli-schwartz commented:
If software is intended to be parallel-installable across major versions, it usually comes with separate dependency names. See for example gtk2 vs. gtk3 vs. gtk4.

Having multiple different providers for the same dependency in the wrapdb might not work 100% fantastically...
